src/linker/Linker/TypeReferenceWalker.cs

```diff
@@ -7,25 +7,26 @@
 using Mono.Cecil.Cil;
 using Mono.Collections.Generic;
 
+#nullable enable
+
 namespace Mono.Linker
 {
-	struct TypeReferenceMarker
+	abstract class TypeReferenceWalker
 	{
-		readonly AssemblyDefinition assembly;
-		readonly MarkingHelpers markingHelpers;
-		HashSet<TypeReference> visited;
+		protected readonly AssemblyDefinition assembly;
+
+		protected HashSet<TypeReference>? visited;
```

Comment:
Maybe turn this into a get-only property which doesn't return null (would throw otherwise, as it should not get called outside of the Process method).
Generally I don't like using the `!` in non-weird cases.

Reply:
Actually - since the usage of this class is always:
```C#
var w = new TypeReferenceWalker(....);
w.Process();
// throw away w
```

Maybe change the class to not allow calling Process multiple times - the derived class would expose a static "ProcessSomething" which internally calls constructor and the Process method - and the state can be initialized in ctor and made readonly.

Reply:
Great ideas, thank you!